cdap-tms/src/test/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberServiceTest.java

```diff
@@ -0,0 +1,242 @@
+ * the License.
+ */
+
+package io.cdap.cdap.messaging.subscriber;
+
+import com.google.common.collect.Iterators;
+import io.cdap.cdap.api.messaging.Message;
+import io.cdap.cdap.api.messaging.MessagingContext;
+import io.cdap.cdap.api.metrics.MetricsContext;
+import io.cdap.cdap.api.metrics.NoopMetricsContext;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.common.utils.ImmutablePair;
+import io.cdap.cdap.proto.id.TopicId;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.transaction.TransactionException;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+public class AbstractMessagingSubscriberServiceTest {
+  @Test
+  public void testMesssagingIterator() throws Exception {
+    int numMessages = 10;
+    List<ImmutablePair<String, String>> messages = createMessages(numMessages);
+    DummySubscriberService<String> dummySubscriber = null;
+
+    // Process full message list.
+    // Verify that the last
+    {
+      dummySubscriber = createDummySubscriber(null,
+                                              iterator -> {
+                                                while (iterator.hasNext()) {
+                                                  iterator.next();
+                                                }
+                                                return null;
+                                              });
+      String lastConsumedId = dummySubscriber.processMessages(Iterators.peekingIterator(messages.iterator()));
+      Assert.assertTrue(lastConsumedId != null);
+      Assert.assertTrue(lastConsumedId.equals(toMessageId(numMessages - 1)));
+      Assert.assertTrue(dummySubscriber.getLastStoredMessageId().equals(toMessageId(numMessages - 1)));
+    }
+
+    // Process partial message list.
+    {
+      final int numProcessed = 5;
+      dummySubscriber = createDummySubscriber(null,
+                                              iterator -> {
+                                                int cnt = 0;
+                                                while (iterator.hasNext()) {
+                                                  iterator.next();
+                                                  cnt++;
+                                                  if (cnt >= numProcessed) {
+                                                    break;
+                                                  }
+                                                }
+                                                return null;
+                                              });
+      String lastConsumedId = dummySubscriber.processMessages(Iterators.peekingIterator(messages.iterator()));
```

Comment:
seems like this is repeated a bunch, could we pull it out into a separate method that takes a dummy subscriber?